### PR TITLE
add multilingual support for origin_info / event mappings

### LINKED ISSUE #1478: Map multilingual origin info
https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/mods_cocina_mappings/mods_to_cocina_originInfo.txt#L1182

See also druid:cp578ch2992

### PR DESCRIPTION
**Pull request by ndushay**
Fixes #1478 

## Why was this change made?

Adds multilingual support for origin_info / event mappings.

Coverage drop is due to no tests for when we hit bizarro MODS and log errors and notify honeybadger.

## How was this change tested?

1. **unit tests** from https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/mods_cocina_mappings/mods_to_cocina_originInfo.txt, including 2 new ones added due to questions as I implemented this.

2. **roundtrip validations**

### Before

```
Status (n=100000):
  Success:   87516 (87.516%)
  Different: 11820 (11.82%)
  To Cocina error:     31 (0.031%)
  To Fedora error:     23 (0.023%)
  Missing:     610 (0.61%)
```

### After

```
Status (n=100000):
  Success:   87907 (87.907%)
  Different: 11429 (11.429%)
  To Cocina error:     31 (0.031%)
  To Fedora error:     23 (0.023%)
  Missing:     610 (0.61%)
```

See comments below for specific to-cocina and to-fedora errors; I am confused by the error counts above not matching those for the separate to-cocina and to-fedora runs ... but whatever.

## Which documentation and/or configurations were updated?

https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/mods_cocina_mappings/mods_to_cocina_originInfo.txt

### CODE REVIEW

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 1]
```diff
@@ -165,6 +314,37 @@ def add_publisher_info(event, set)
...
+
+          publication_event = events.find { |event| event[:type] == 'publication' }
+
+          orig_contributors = publication_event[:contributor]
+          orig_contrib_name_value = first_contrib_name_value(orig_contributors)
+          return nil unless orig_contrib_name_value
+
+          parallel_value = parallel_value(orig_contrib_name_value, parallel_contrib_value, scripts_langs_hash)
+          if orig_contributors.size > 1
+            additional_values = orig_contributors.select do |contrib|
```
Comment (aaron-collier) [inline app/services/cocina/from_fedora/descriptive/event.rb:329]: Looks like this branch did not get test coverage.

Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/event.rb:329]: thanks!  I was trying to find that!

Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/event.rb]: this code was to trap for an obscure case that isn't in our specs, so I removed it.

**Comment by ndushay** [comment]
@aaron-collier - turns out the untested code you flagged wasn't really necessary, so I took it out;  I think the remaining coverage dip is due to trapping for unexpected MODS coditions in origin info and logging errors and notifying Honeybadger.

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 2]
```diff
@@ -88,18 +173,68 @@ def add_place_info(event, place_set)
...
+          orig_locations = event[:location]
+          orig_location_value = first_value(orig_locations)
+          return nil unless orig_location_value
+
+          parallel_value = parallel_value(orig_location_value, parallel_place_value, scripts_langs_hash)
+          if orig_locations.size > 1
+            additional_values = orig_locations.select { |location| location[:value].present? && location[:value] != orig_location_value }
+            parallel_value = add_to_parallel_value(parallel_value, additional_values) if additional_values.present?
+          end
+          orig_w_value = first_with_value(orig_locations)
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/event.rb:190]: Can you use a more descriptive variable than `orig_w_value`?

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 3]
```diff
@@ -88,18 +173,68 @@ def add_place_info(event, place_set)
...
           cocina[:uri] = xml_node['valueURI'] if xml_node['valueURI']
           if xml_node['authority']
             cocina[:source] = {
               code: Authority.normalize_code(xml_node['authority']),
               uri: Authority.normalize_uri(xml_node['authorityURI'])
             }.compact
+          elsif xml_node['authorityURI'] # used for originInfo placeTermn
+            cocina[:source] = {
+              uri: Authority.normalize_uri(xml_node['authorityURI'])
+            }.compact
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/event.rb:237]: Is the `if ... elsif` necessary? Won't it the `.compact` clean up the null?

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 4]
```diff
@@ -165,6 +320,31 @@ def add_publisher_info(event, set)
...
+
+          parallel_value = parallel_value(orig_contrib_name_value, parallel_contrib_value, scripts_langs_hash)
+          publication_event[:contributor].first[:name] = [parallel_value]
+
+          addl_contributors = orig_contributors.reject { |contrib| contrib[:name].present? }
+          addl_contributors.each { |contrib_val| publication_event[:contributor] << contrib_val }
+        end
+
+        def first_contrib_name_value(orig_contributors)
+          orig_contrib_name_w_value = orig_contributors&.find do |contrib|
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/event.rb:341]: Can you use a more descriptive variable name than `orig_contrib_name_w_value`? Not clear what the `_w_` is.

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 5]
```diff
@@ -214,6 +413,45 @@ def date_other_event_type(origin)
         def origin_info
           @origin_info ||= resource_element.xpath(ORIGININFO_XPATH, mods: DESC_METADATA_NS)
         end
+
+        def parallel_value(orig_value, parallel_value, scripts_langs_hash)
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/event.rb:417]: Can't there by more that 2 parallel values?

File: app/services/cocina/from_fedora/descriptive/titles.rb [hunk 6]
```diff
@@ -37,7 +37,7 @@ def build(require_title: true)
           title_infos_with_groups = resource_element.xpath('mods:titleInfo[@altRepGroup]', mods: DESC_METADATA_NS)
           grouped_title_infos = title_infos_with_groups.group_by { |node| node['altRepGroup'] }
 
-          result = grouped_title_infos.map { |_k, node_set| parallel(node_set) }
+          result = grouped_title_infos.map { |_k, node_set_array| parallel(node_set_array) }
```
Comment (justinlittman) [inline app/services/cocina/from_fedora/descriptive/titles.rb:40]: In XML terminology, a node set is an array. Why the change?

**Review by justinlittman — COMMENTED** [review]
Holey moley.

File: app/services/cocina/from_fedora/descriptive/titles.rb (hunk 6 above)
Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/titles.rb:40]: it's an array of node_sets, IIRC.

File: app/services/cocina/from_fedora/descriptive/event.rb (hunk 3 above)
Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/event.rb:237]:
```
source = { code: xx, uri: xx }
cocina[:source] = source if source.present?
```

File: app/services/cocina/from_fedora/descriptive/event.rb [hunk 7]
```diff
@@ -20,18 +20,107 @@ def initialize(resource_element:)
...
+          node_set_array.each do |origin_info_node_set|
+            parallel_attribs = origin_info_node_set&.attributes
+            scripts_langs_hash[:parallel_script] = parallel_attribs['script']&.value
+            scripts_langs_hash[:parallel_lang_code] = parallel_attribs['lang']&.value
+            origin_info_node_set.element_children.each do |child_el|
+              child_el_name = child_el.name
+              if el_names_in_common.include?(child_el_name)
+                build_parallel_value(events, child_el, event_type, scripts_langs_hash)
+              else
+                errmsg = "problem building event parallel values due to unmatched originInfo element #{child_el_name}"
```
[the remark this answers is not in the document]
Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/event.rb:77]: done.

File: app/services/cocina/from_fedora/descriptive/titles.rb (hunk 6 above)
Comment (ndushay) [reply app/services/cocina/from_fedora/descriptive/titles.rb]: I backed this out.  NodeSet implements Enumerable so it shows up as an Array in pry ... confusing to me.

**Comment by ndushay** [comment]
## Roundtrip

### Before

```
Status (n=250000):
  Success:   215062 (86.0248%)
  Different: 33338 (13.3352%)
  To Cocina error:     72 (0.0288%)
  To Fedora error:     59 (0.0236%)
  Missing:     1469 (0.5876%)
```

## To Cocina

### Before

```
```

## To Fedora

### Before

```
```

**Comment by ndushay** [comment]
## Roundtrip

### Before

```
Status (n=250000):
  Success:   218563 (87.4252%)
  Different: 29839 (11.9356%)
  To Cocina error:     70 (0.028%)
  To Fedora error:     59 (0.0236%)
  Missing:     1469 (0.5876%)
```

### After

```
Status (n=250000):
  Success:   219537 (87.8148%)
  Different: 28510 (11.404%)
  To Cocina error:     125 (0.05%)
  To Fedora error:     359 (0.1436%)
  Missing:     1469 (0.5876%)
```

## To Cocina

### Before

```
To Cocina error: 4 of 250000 (0.0016%)
Data error: 31461 of 250000 (12.5844%)
Missing: 1469 of 250000 (0.5876%)


```

### After

## To Fedora

### Before

```

```

### After

**Comment by ndushay** [comment]
## To Fedora

### Before

```

```

### After

```

```

**Comment by justinlittman** [comment]
Before:
```
Status (n=100000):
  Success:   87731 (87.731%)
  Different: 11607 (11.607%)
  To Cocina error:     31 (0.031%)
  To Fedora error:     21 (0.021%)
  Missing:     610 (0.61%)
```

After:
```
Status (n=100000):
  Success:   88122 (88.122%)
  Different: 11216 (11.216%)
  To Cocina error:     31 (0.031%)
  To Fedora error:     21 (0.021%)
  Missing:     610 (0.61%)
```

**Review by justinlittman — APPROVED** [review]
